plugin/pkg/shoot/validator/admission.go

```diff
@@ -448,6 +449,24 @@ func (c *validationContext) validateScheduling(ctx context.Context, a admission.
 		}
 	}
 
+	if mustCheckSchedulingConstraints {
+		if seedSelector := c.cloudProfile.Spec.SeedSelector; seedSelector != nil {
+			selector, err := metav1.LabelSelectorAsSelector(&seedSelector.LabelSelector)
+			if err != nil {
+				return apierrors.NewInternalError(fmt.Errorf("label selector conversion failed: %v for seedSelector: %w", seedSelector.LabelSelector, err))
+			}
+			if !selector.Matches(labels.Set(c.seed.Labels)) {
+				return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' because the seed selector of cloud profile '%s' is not matching the labels of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name))
+			}
+
+			if seedSelector.ProviderTypes != nil {
+				if !sets.NewString(seedSelector.ProviderTypes...).HasAny(c.seed.Spec.Provider.Type, "*") {
+					return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' because none of the provider types in the seed selector of cloud profile '%s' is matching the provider type of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name))
+				}
+			}
```

Comment:
Can we add these to an `ErrorList` and return instead of returning at the first occurence? Both of them can happen parallely.

Reply:
You are right that both the the errors can occur at the same time.

If we decide to use the `ErrorList` approach only for the newly introduced checks, it would look like:
```diff
@@ -455,15 +455,26 @@ func (c *validationContext) validateScheduling(ctx context.Context, a admission.
 			if err != nil {
 				return apierrors.NewInternalError(fmt.Errorf("label selector conversion failed: %v for seedSelector: %w", seedSelector.LabelSelector, err))
 			}
+
+			var (
+				allErrs field.ErrorList
+				path    = field.NewPath("spec", "seedName")
+			)
 			if !selector.Matches(labels.Set(c.seed.Labels)) {
-				return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' because the seed selector of cloud profile '%s' is not matching the labels of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name))
+				detail := fmt.Sprintf("cannot schedule shoot '%s' on seed '%s' because the seed selector of cloud profile '%s' is not matching the labels of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name)
+				allErrs = append(allErrs, field.Invalid(path, c.shoot.Spec.SeedName, detail))
 			}

 			if seedSelector.ProviderTypes != nil {
 				if !sets.NewString(seedSelector.ProviderTypes...).HasAny(c.seed.Spec.Provider.Type, "*") {
-					return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' because none of the provider types in the seed selector of cloud profile '%s' is matching the provider type of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name))
+					detail := fmt.Sprintf("cannot schedule shoot '%s' on seed '%s' because none of the provider types in the seed selector of cloud profile '%s' is matching the provider type of the seed", c.shoot.Name, c.seed.Name, c.cloudProfile.Name)
+					allErrs = append(allErrs, field.Invalid(path, c.shoot.Spec.SeedName, detail))
 				}
 			}
+
+			if len(allErrs) > 0 {
+				return admission.NewForbidden(a, fmt.Errorf("%+v", allErrs))
+			}
 		}
 	}
```

The new checks would be inconsistent with the other ones in the `validateScheduling` func. I rather think that the `validateScheduling` func has to be adapted to return `ErrorList`. I would vote to keep the current checks as they are for the score of this PR. If you want, I can try to prepare a new PR that adapts the `validateScheduling` func to return `ErrorList`. WDYT?

Reply:
Yes, I noticed that as well. You can keep it like this, I have already prepared a PR on top of this to cleanup binding plugin, I'll make the changes.

Reply:
WIP commit that I prepared locally - https://github.com/ialidzhikov/gardener/commit/cb724fc01798d9983af6c5a86eb9126eefafa034

Reply:
Then please go ahead and open the PR. :)